packages/peregrine/lib/store/actions/cart/asyncActions.js

```diff
@@ -360,6 +360,50 @@ export const getCartDetails = payload => {
     };
 };
 
+/**
+ * Merges Guest cart with Logged in user's cart
+ *
+ * @param payload.mergeCarts {Object} merge carts mutation hook object
+ * @param payload.fetchCartId {Object} fetch cart id mutation hook object
+ */
+export const retrieveAndMergeCarts = payload =>
+    async function thunk(dispatch) {
+        const { mergeCarts, fetchCartId } = payload;
+        const sourceCartId = await retrieveCartId();
+
+        try {
+            // remove guest cart id from local storage
+            await dispatch(removeCart());
+            // create customer cart
+            await dispatch(
+                createCart({
+                    fetchCartId
+                })
+            );
```

Comment:
We might be able to do away with the async action for merging if in the sign in callback we:

1) Obtain a reference to guest cart id or "source id"
2) Clear all cart data from cache/redux
3) Query for `customerCart` to get "destination id"
4) Merge carts. Spread response through cache/redux.

This logic is basically what was written here in the action, but since we are going to be moving towards removing much of the cart slice it makes sense to try to do this in the talons. Moving it to the talon also allows us to directly call `clearCartDataFromCache` at the same time we call the `removeCart` action.

Reply:
@sirugh how to dispatch in talons, if any rewrite of async action to talon is done, can you share that?.

Reply:
> how to dispatch in talons, if any rewrite of async action to talon is done, can you share that?.

Of course I can share, but you've already done this by calling your new `retrieveAndMerge` function from the talon. I am just proposing that we just call `removeCart` from the talon rather than from the action.

Once we get data back from `mergeCart` we can then also pass that to an action for storing in redux.

Since we are planning on deprecating the old mini cart logic, most of the old actions will get obsoleted anyways so I want to avoid adding new logic there if possible.